substrate/client/runtime-utilities/src/lib.rs

```diff
@@ -48,21 +57,24 @@ pub fn fetch_latest_metadata_from_code_blob<HF: HostFunctions>(
 				.call("Metadata_metadata_at_version", latest_stable)
 				.map_err(|_| "Unable to fetch metadata from blob".to_string())?;
 
-			Option::<OpaqueMetadata>::decode(&mut encoded.as_slice())?
-				.ok_or_else(|| "Metadata not found".to_string())?
+			Option::<OpaqueMetadata>::decode(&mut encoded.as_slice())
+				.map_err(|_| <&str as Into<Error>>::into("Unable to decode to opaque metadata"))?
+				.ok_or_else(|| "Opaque metadata not found".to_string())?
 		},
 		Err(_) => {
 			let encoded = runtime_caller
 				.call("Metadata_metadata", ())
 				.map_err(|_| "Unable to fetch metadata from blob".to_string())?;
-			Decode::decode(&mut encoded.as_slice())?
+			Decode::decode(&mut encoded.as_slice())
+				.map_err(|_| <&str as Into<Error>>::into("Unable to decode to opaque metadata"))?
 		},
 	};
 
-	Ok(subxt::Metadata::decode(&mut (*opaque_metadata).as_slice())?)
+	subxt::Metadata::decode(&mut (*opaque_metadata).as_slice())
+		.map_err(|_| "Unable to decode opaque metadata".into())
 }
 
-struct BasicCodeFetcher<'a> {
+pub struct BasicCodeFetcher<'a> {
```

Comment:
nit: if this is public, doc would be nice.

Reply:
I thought again it could be used independently, but it is used just in the crate for now. I think similar logic can be exposed publicly from somewhere (not sure if this is the best place though) and used everywhere runtime calls are made via the executor, but I wouldn't tackle right now making such an API public (and use it everywhere `FetchRuntimeCode` trait is implemented and used in the same way).

Removed the pub here: [0ba6ecb](https://github.com/paritytech/polkadot-sdk/pull/6450/commits/0ba6ecb244c1ae43bfc6262e2c1143bf3f735410)